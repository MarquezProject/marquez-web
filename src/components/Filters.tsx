import React, { ReactElement, useState } from 'react'
import MUISelect from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'
import Box from '@material-ui/core/Box'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'
import uniq from 'lodash/uniq'
import { withStyles } from '@material-ui/core/styles'
import { capitalize } from '../helpers'
import { IProps } from '../containers/FilterContainer'
// import { IFilterByDisplay, IFilterByKey } from '../types'

const StyledFormControl = withStyles({
  root: {
    margin: '0rem 2rem',
    minWidth: '10rem'
  }
})(FormControl)

interface IFilterDictionary {
  [key: string]: {
    entities: any[]
    accessor: (n: any) => any
  }
}

const filterByOptions: { [key: string]: 'namespace' | 'sourceName' } = {
  namespace: 'namespace',
  datasource: 'sourceName'
}

const Filters = (props: IProps): ReactElement => {
  const { namespaces, datasets, filterJobs, filterDatasets, showJobs } = props
  const [currentFilter, setCurrentFilter] = useState('all')

  const [showSubFilter, toggleShowSubFilter] = useState(false)

  const datasources = uniq(datasets.map(d => d.sourceName))

  const filterDictionary: IFilterDictionary = {
    namespace: {
      entities: namespaces,
      accessor: n => n.name
    },
    datasource: {
      entities: datasources,
      accessor: n => n
    }
  }

  const [currentFilterValue, setCurrentFilterValue] = useState({})
  return (
    <Box p={2}>
      <StyledFormControl margin='normal'>
        <InputLabel id='filter-by-label'>Filter by</InputLabel>
        <MUISelect
          value={currentFilter}
          renderValue={capitalize}
          onChange={e => {
            const newValue = e.target.value as 'namespace' | 'datasource' | 'all'
            // debugger
            setCurrentFilter(newValue)
            if (newValue === 'all') {
              toggleShowSubFilter(false)
            } else {
              setCurrentFilterValue(
                filterDictionary[e.target.value as 'namespace' | 'datasource'].entities[0]
              )
              toggleShowSubFilter(true)
            }
          }}
        >
          {Object.keys(filterByOptions).map(o => (
            <MenuItem key={o} value={o}>
              {o}
            </MenuItem>
          ))}
          <MenuItem key='all' value='all'>
            All
          </MenuItem>
        </MUISelect>
      </StyledFormControl>
      {showSubFilter && (
        <StyledFormControl margin='normal'>
          <InputLabel id='secondary-filter'>{currentFilter}</InputLabel>
          <MUISelect
            value={currentFilterValue}
            renderValue={filterDictionary[currentFilter].accessor}
            onChange={e => {
              const currentFilterValue = e.target.value as string
              const currentFilterKeyDisplay = filterByOptions[currentFilter]
              const currentFilterKey = filterByOptions[currentFilterKeyDisplay]
              setCurrentFilterValue(currentFilterValue)
              filterJobs(
                currentFilterKey,
                filterDictionary[currentFilter].accessor(currentFilterValue)
              )
              filterDatasets(
                currentFilterKey,
                filterDictionary[currentFilter].accessor(currentFilterValue)
              )
              showJobs(true)
            }}
          >
            {filterDictionary[currentFilter].entities.map(o => {
              const val = filterDictionary[currentFilter].accessor(o)
              return (
                <MenuItem key={val} value={o}>
                  {val}
                </MenuItem>
              )
            })}
          </MUISelect>
        </StyledFormControl>
      )}
    </Box>
  )
}

export default Filters
