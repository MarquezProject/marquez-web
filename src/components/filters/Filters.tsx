import { IProps } from './FiltersWrapper'
import { Namespace } from '../../types/api'
import { capitalize } from '../../helpers'
import { withStyles } from '@material-ui/core/styles'
import Box from '@material-ui/core/Box'
import FormControl from '@material-ui/core/FormControl'
import MUISelect from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'
import React, { ChangeEvent, ReactElement, useState } from 'react'
import uniq from 'lodash/uniq'

const StyledFormControl = withStyles({
  root: {
    margin: '0rem 2rem 0rem 0rem',
    minWidth: '10rem'
  }
})(FormControl)

type IEntity = Namespace | string
interface IFilterDictionary {
  [key: string]: {
    entities: IEntity[]
    accessor: (n: IEntity) => string
    default: any
  }
}

const filterByOptions: { [key: string]: 'namespace' | 'sourceName' } = {
  /* display name: entitiy name on dataset schema */
  namespace: 'namespace',
  source: 'sourceName'
}

const Filters = (props: IProps): ReactElement => {
  const { namespaces, datasets, filterJobs, filterDatasets, showJobs } = props

  const [currentFilter, setCurrentFilter] = useState('all')
  const [currentFilterValue, setCurrentFilterValue] = useState({})
  const [subFilterVisible, setSubFilterVisibility] = useState(false)

  const sources = uniq(datasets.map(d => d.sourceName))
  const filterDictionary: IFilterDictionary = {
    namespace: {
      entities: namespaces,
      accessor: n => (n as Namespace).name,
      default: { name: '' }
    },
    source: {
      entities: sources,
      accessor: (n: string): string => n,
      default: ''
    }
    /* Can add more filters here */
  }

  /* Set the category we will be filtering by */
  const onPrimaryFilterChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const newPrimaryFilter = e.target.value as 'namespace' | 'source' | 'all'
    setCurrentFilter(newPrimaryFilter)
    if (newPrimaryFilter === 'all') {
      setSubFilterVisibility(false)
      filterJobs(newPrimaryFilter)
      filterDatasets(newPrimaryFilter)
    } else {
      const currentFilterConfig = filterDictionary[newPrimaryFilter]
      const currentFilterValue = currentFilterConfig.entities[0] || currentFilterConfig.default
      const currentFilterKey = filterByOptions[newPrimaryFilter]
      setCurrentFilterValue(currentFilterValue)
      setSubFilterVisibility(true)
      filterJobs(currentFilterKey, currentFilterConfig.accessor(currentFilterValue))
      filterDatasets(currentFilterKey, currentFilterConfig.accessor(currentFilterValue))
    }
    showJobs(true)
  }

  /* Filter jobs & datasets by the selected filter value */
  const onSecondaryFilterChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const currentFilterValue = e.target.value as string
    const currentFilterKey = filterByOptions[currentFilter]

    setCurrentFilterValue(currentFilterValue)
    filterJobs(currentFilterKey, filterDictionary[currentFilter].accessor(currentFilterValue))
    filterDatasets(currentFilterKey, filterDictionary[currentFilter].accessor(currentFilterValue))
    showJobs(true)
  }

  return (
    <Box py={2}>
      <StyledFormControl margin='normal'>
        <MUISelect value={currentFilter} renderValue={capitalize} onChange={onPrimaryFilterChange}>
          {Object.keys(filterByOptions).map(o => (
            <MenuItem key={o} value={o}>
              {o}
            </MenuItem>
          ))}
          <MenuItem key='all' value='all'>
            all
          </MenuItem>
        </MUISelect>
      </StyledFormControl>
      {subFilterVisible && (
        <StyledFormControl margin='normal'>
          <MUISelect
            value={currentFilterValue}
            renderValue={filterDictionary[currentFilter].accessor}
            onChange={onSecondaryFilterChange}
          >
            {filterDictionary[currentFilter].entities.map((o: any) => {
              const val = filterDictionary[currentFilter].accessor(o)
              return (
                <MenuItem key={val} value={o}>
                  {val}
                </MenuItem>
              )
            })}
          </MUISelect>
        </StyledFormControl>
      )}
    </Box>
  )
}

export default Filters
