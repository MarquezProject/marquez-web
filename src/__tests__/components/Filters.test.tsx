import { shallow } from 'enzyme'
import * as React from 'react'
import Filters from '../../components/filters/Filters'
import Select from '@material-ui/core/Select'
const namespaces = require('../../../docker/db/data/namespaces.json')

describe('Filters component', () => {
  const mockProps = {
    datasets: [],
    jobs: [],
    namespaces,
    filterDatasets: () => {},
    filterJobs: () => {}
  }

  const wrapper = shallow(<Filters {...mockProps} />)

  it('should render only one Select component on initial render', () => {
    expect(wrapper.find(Select)).toHaveLength(1)
  })

  test.skip('should render second Select component after onChange event for first Select', () => {
    const clickEvent = { target: { value: 'namespace' } }
    wrapper.find(Select).simulate('change', clickEvent)
    expect(wrapper.find(Select)).toHaveLength(2)
  })

  it('renders a snapshot that matches previous', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
