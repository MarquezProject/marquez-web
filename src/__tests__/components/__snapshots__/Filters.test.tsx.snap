// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component renders a snapshot that matches previous 1`] = `
<Styled(MuiBox)
  ml="5%"
  py={2}
>
  <WithStyles(WithStyles(ForwardRef(FormControl)))
    margin="normal"
  >
    <WithStyles(ForwardRef(Select))
      onChange={[Function]}
      renderValue={[Function]}
      value="all"
    >
      <WithStyles(ForwardRef(MenuItem))
        key="namespace"
        value="namespace"
      >
        namespace
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        key="datasource"
        value="datasource"
      >
        datasource
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        key="all"
        value="all"
      >
        all
      </WithStyles(ForwardRef(MenuItem))>
    </WithStyles(ForwardRef(Select))>
  </WithStyles(WithStyles(ForwardRef(FormControl)))>
</Styled(MuiBox)>
`;
